ssl/ssl_cert_comp.c

```diff
@@ -0,0 +1,464 @@
+{
+    size_t ret;
+    /*
+     * Uncompressibility expansion:
+     * ZLIB: N + 11 + 5 * (N >> 14)
+     * Brotli: per RFC7932: N + 5 + 3 * (N >> 16)
+     * ZSTD: N + 4 + 14 + 3 * (N >> 17) + 4
+     */
+    
+    switch (alg) {
+    case TLSEXT_comp_cert_zlib:
+        ret = length + 11 + 5 * (length >> 14);
+        break;
+    case TLSEXT_comp_cert_brotli:
+        ret = length + 5 + 3 * (length >> 16);
+        break;
+    case TLSEXT_comp_cert_zstd:
+        ret = length + 22 + 3 * (length >> 17);
+        break;
+    default:
+        return 0;
+    }
+    /* Check for overflow */
+    if (ret < length)
+        return 0;
+    return ret;
+}
+
+int ossl_comp_has_alg(int a)
+{
+#ifndef OPENSSL_NO_COMP_ALG
+    /* 0 means "any" algorithm */
+    if ((a == 0 || a == TLSEXT_comp_cert_brotli) && BIO_f_brotli() != NULL)
+        return 1;
+    if ((a == 0 || a == TLSEXT_comp_cert_zstd) && BIO_f_zstd() != NULL)
+        return 1;
+    if ((a == 0 || a == TLSEXT_comp_cert_zlib) && BIO_f_zlib() != NULL)
+        return 1;
+#endif
+    return 0;
+}
+
+/* New operation Helper routine */
+#ifndef OPENSSL_NO_COMP_ALG
+static OSSL_COMP_CERT *OSSL_COMP_CERT_new(unsigned char *data, size_t len, size_t orig_len, int alg)
+{
+    OSSL_COMP_CERT *ret = NULL;
+
+    if (!ossl_comp_has_alg(alg)
+            || data == NULL
+            || (ret = OPENSSL_zalloc(sizeof(*ret))) == NULL
+            || (ret->lock = CRYPTO_THREAD_lock_new()) == NULL)
+        goto err;
+
+    ret->references = 1;
+    ret->data = data;
+    ret->len = len;
+    ret->orig_len = orig_len;
+    ret->alg = alg;
+    return ret;
+ err:
+    ERR_raise(ERR_LIB_SSL, ERR_R_MALLOC_FAILURE);
+    OPENSSL_free(data);
+    OPENSSL_free(ret);
+    return NULL;
+}
+
+__owur static OSSL_COMP_CERT *OSSL_COMP_CERT_compressed_data(unsigned char *data, size_t len, size_t orig_len, int alg)
+{
+    return OSSL_COMP_CERT_new(OPENSSL_memdup(data, len), len, orig_len, alg);
+}
+
+__owur static OSSL_COMP_CERT *OSSL_COMP_CERT_uncompressed_data(unsigned char *data, size_t len, int alg)
+{
+    OSSL_COMP_CERT *ret = NULL;
+    size_t max_length;
+    int comp_length;
+    COMP_METHOD *method;
+    unsigned char *comp_data = NULL;
+    COMP_CTX *comp_ctx = NULL;
+
+    switch (alg) {
+    case TLSEXT_comp_cert_brotli:
+        method = COMP_brotli_oneshot();
+        break;
+    case TLSEXT_comp_cert_zlib:
+        method = COMP_zlib();
+        break;
+    case TLSEXT_comp_cert_zstd:
+        method = COMP_zstd_oneshot();
+        break;
+    default:
+        goto err;
+    }
+
+    if ((max_length = ossl_calculate_comp_expansion(alg, len)) == 0
+          || method == NULL
+          || (comp_ctx = COMP_CTX_new(method)) == NULL
+          || (comp_data = OPENSSL_zalloc(max_length)) == NULL)
+        goto err;
+
+    comp_length = COMP_compress_block(comp_ctx, comp_data, max_length, data, len);
+    if (comp_length <= 0)
+        goto err;
+
+    ret = OSSL_COMP_CERT_new(comp_data, comp_length, len, alg);
```

Comment:
`comp_data` buffer is allocated, we compress the data into it, then `OSSL_COMP_CERT_new` dups it, and then we free the original buffer. Seems like quite a bit of churn. Is the dup really necessary?

Reply:
`OSSL_COMP_CERT_new()` does not duplicate `comp_data`, it takes ownership of it. So I'm unsure what you're referring to here?